core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -93,90 +93,113 @@ class TransactionCoordinator(brokerId: Int,
+          val coordinatorEpoch = epochAndMetadata.coordinatorEpoch
+          val txnMetadata = epochAndMetadata.transactionMetadata
+
+          // there might be a concurrent thread that has just updated the mapping
+          // with the transactional id at the same time (hence reference equality will fail);
+          // in this case we will treat it as the metadata has existed already
+          txnMetadata synchronized {
+            if (!txnMetadata.eq(createdMetadata)) {
+              initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, coordinatorEpoch, txnMetadata)
+            } else {
+              Right(coordinatorEpoch, txnMetadata.prepareNewPid(time.milliseconds()))
             }
-          case Some(metadata) =>
-            initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
-        }
+          }
+
+        case Some(existingEpochAndMetadata) =>
+          val coordinatorEpoch = existingEpochAndMetadata.coordinatorEpoch
+          val txnMetadata = existingEpochAndMetadata.transactionMetadata
+
+          txnMetadata synchronized {
+            initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, coordinatorEpoch, txnMetadata)
+          }
       }
-  }
 
-  private def appendMetadataToLog(transactionalId: String,
-                             metadata: TransactionMetadata,
-                             initPidCallback: InitPidCallback): Unit ={
-    def callback(errors: Errors): Unit = {
-      if (errors == Errors.NONE)
-        initPidCallback(initTransactionMetadata(metadata))
-      else
-        initPidCallback(initTransactionError(errors))
+      result match {
+        case Left(pidResult) =>
+          responseCallback(pidResult)
+
+        case Right((coordinatorEpoch, newMetadata)) =>
+          if (newMetadata.txnState == Ongoing) {
+            // abort the ongoing transaction and then return CONCURRENT_TRANSACTIONS to let client wait and retry
+            def sendRetriableErrorCallback(error: Errors): Unit = {
+              if (error != Errors.NONE) {
+                responseCallback(initTransactionError(error))
+              } else {
+                responseCallback(initTransactionError(Errors.CONCURRENT_TRANSACTIONS))
+              }
+            }
+
+            handleEndTransaction(transactionalId,
+              newMetadata.producerId,
+              newMetadata.producerEpoch,
+              TransactionResult.ABORT,
+              sendRetriableErrorCallback)
+          } else {
+            def sendPidResponseCallback(error: Errors): Unit = {
+              if (error == Errors.NONE)
+                responseCallback(initTransactionMetadata(newMetadata))
+              else
+                responseCallback(initTransactionError(error))
+            }
+
+            txnManager.appendTransactionToLog(transactionalId, coordinatorEpoch, newMetadata, sendPidResponseCallback)
```

Comment:
Yup, it is going to be done in the locking PR that I'm also working on now. Just trying to keep each PR small to make it easier for reviews.